### PR TITLE
Added check to verify that supply is not exceeded

### LINKED ISSUE #78: Fix: Seed.sol does not guarantee a sufficient supply of seed tokens
When a buyer claims their tokens once the minimum threshold is reached, there is no guarantee there is enough `seedToken` in the contract to pay them. The call to`claimLock` will revert because of the lack of `seedTokens`. They cannot rescue their tokens using `buyBack` either after the threshold is reached. Effectively,the user’s funding tokens are stuck in the contract.

`TODO`:

- Add require, in the `buy` function, that there are enough `seedTokens` available for the user to buy (as well as to pay the fees)

### PR DESCRIPTION
**Pull request by BritikovKI**
Closes #78 
Review calculations pls, not sure about correctness(especially in division part)

### CODE REVIEW

File: contracts/seed/Seed.sol [hunk 1]
```diff
@@ -159,6 +159,9 @@ contract Seed {
     */
     function buy(uint256 _amount) public protected checked {
         require((fundingToken.balanceOf(address(this)).add(_amount)) <= cap, "Seed: amount exceeds contract sale cap");
+        require( seedToken.balanceOf(address(this)) >=
+        (fundingToken.balanceOf(address(this)).add(_amount)).mul(price).div(PCT_BASE),
+            "Seed: seed distribution exceeded");
         require(fundingToken.transferFrom(msg.sender, address(this), _amount), "Seed: no tokens");
```
Comment (kinspotting) [inline contracts/seed/Seed.sol:165]: pls transfer `_amount.mul(price)`

File: contracts/seed/Seed.sol [hunk 2]
```diff
@@ -159,6 +159,9 @@ contract Seed {
     */
```
Comment (kinspotting) [inline contracts/seed/Seed.sol:159]: pls change comment to  `The amount of seed tokens to buy.`

Comment (kinspotting) [reply contracts/seed/Seed.sol:159]: + in Lock structure pls change `amount` to `seedAmount`

File: contracts/seed/Seed.sol (hunk 1 above)
Comment (kinspotting) [reply contracts/seed/Seed.sol:165]: In `172: _addLock(msg.sender, (_lockTokens.add((_amount).mul(price).div(PCT_BASE))), _amount, feeAmount);` pls adjust `seedAmount` and `fundingAmount`

**Review by kinspotting — CHANGES_REQUESTED** [review]
needs changes

File: contracts/seed/Seed.sol (hunk 1 above)
Comment (BritikovKI) [reply contracts/seed/Seed.sol]: Done

File: contracts/seed/Seed.sol (hunk 2 above)
Comment (BritikovKI) [reply contracts/seed/Seed.sol:159]: Done

File: test/seed.js [hunk 3]
```diff
@@ -259,16 +264,20 @@ contract('Seed', (accounts) => {
...
                 });
             });
             context('» getAmount', () => {
                 it('returns correct amount', async () => {
-                    let pct_base = new BN('1000000000000000000'); // 10**18
-                    let p = new BN(price);
-                    let a = new BN(buyAmount);
-                    let amount = new BN(p.mul(a).div(pct_base));
-                    expect((await setup.seed.getAmount(buyer1)).toString()).to.equal((amount).toString());
+                    // TODO: all this numbers doesn't make sense anymore, as we are sending amount of seedTokens we
```
Comment (kinspotting) [inline test/seed.js:275]: please delete the comment

File: contracts/seed/Seed.sol [hunk 4]
```diff
@@ -155,20 +155,26 @@ contract Seed {
...
-      * @param _amount           The amount of tokens to buy.
+      * @param _seedAmount       The amount of seed tokens to buy.
     */
-    function buy(uint256 _amount) public protected checked {
-        require((fundingToken.balanceOf(address(this)).add(_amount)) <= cap, "Seed: amount exceeds contract sale cap");
-        require(fundingToken.transferFrom(msg.sender, address(this), _amount), "Seed: no tokens");
+    function buy(uint256 _seedAmount) public protected checked {
+        //        TODO: DEAR ZNATOKI, Attention, question from our viewers, why do we need to divide anything anyway
+        uint256 fundingAmount = (_seedAmount.mul(price)).div(PCT_BASE);
+        require((fundingToken.balanceOf(address(this)).add(fundingAmount)) <= cap,
```
Comment (kinspotting) [inline contracts/seed/Seed.sol:163]: please include fee to the check

**Review by kinspotting — CHANGES_REQUESTED** [review]
needs changes

File: contracts/seed/Seed.sol [hunk 5]
```diff
@@ -155,20 +155,26 @@ contract Seed {
...
     /**
       * @dev                     Buy seed tokens.
-      * @param _amount           The amount of tokens to buy.
+      * @param _seedAmount       The amount of seed tokens to buy.
     */
-    function buy(uint256 _amount) public protected checked {
-        require((fundingToken.balanceOf(address(this)).add(_amount)) <= cap, "Seed: amount exceeds contract sale cap");
-        require(fundingToken.transferFrom(msg.sender, address(this), _amount), "Seed: no tokens");
+    function buy(uint256 _seedAmount) public protected checked {
+        //        TODO: DEAR ZNATOKI, Attention, question from our viewers, why do we need to divide anything anyway
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply contracts/seed/Seed.sol:161]: Roger Roger

File: contracts/seed/Seed.sol (hunk 4 above)
Comment (BritikovKI) [reply contracts/seed/Seed.sol:163]: Included, but actually have one more question, I think we do need to include price of the fee tokens to overall price in the fundingTokens, that we are sending to the contract?

Comment (BritikovKI) [reply contracts/seed/Seed.sol:163]: And why are we paying fee in SeedTokens anyways and not in the FundingTokens?

File: contracts/seed/Seed.sol [hunk 6]
```diff
@@ -196,11 +220,12 @@ contract Seed {
...
         Lock storage tokenLock = tokenLocks[msg.sender];
         uint amount = tokenLock.fundingAmount;
-        tokenLock.amount = 0;
+        tokenLock.seedAmount = 0;
         tokenLock.fee = 0;
         tokenLock.fundingAmount = 0;
+        uint feeAmount = (amount.mul(uint(PPM))).mul(fee).div(PPM100);
         require(
-            fundingToken.transfer(msg.sender, amount),
+            fundingToken.transfer(msg.sender, (amount+feeAmount)),
```
Comment (kinspotting) [inline contracts/seed/Seed.sol:228]: please use SafeMath

**Review by kinspotting — CHANGES_REQUESTED** [review]
needs a minor change

File: contracts/seed/Seed.sol (hunk 6 above)
Comment (BritikovKI) [reply contracts/seed/Seed.sol]: Done

**Review by kinspotting — APPROVED** [review]
lgtm

**Pull request merged**